### PR TITLE
[ASTS] Implementation of SweepBucketAssignerStateMachineTable

### PR DESCRIPTION
**Pull request by mdaudali**
## General
**Before this PR**:
Breaking up https://github.com/palantir/atlasdb/pull/7313 into individual PRs representing each table. They will all be in the same class.

We have various interfaces representing logical tables necessary for tracking bucket state, but no concrete implementation.
**After this PR**:
Start of an initial implementation.

==COMMIT_MSG==
==COMMIT_MSG==

**Priority**: P2

**Concerns / possible downsides (what feedback would you like?)**:
Check the setInitialStateForBucketAssigner, which is new.

Also, I've done a drive by to the client smile mapper, from the server mapper. The server mappers enable FAIL_ON_UNKNOWN_PROPERTIES, which will prevent us from adding optional fields in the future. Do check this.
## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:
N/A
**What was existing testing like? What have you done to improve it?**:
Added tests
**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
N/A

## Execution
Not wired up.

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
No
**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
No
**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
Not this one
## Development Process
**Where should we start reviewing?**:
DBA
**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:
N/A
**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Override
+    public void updateStateMachineForBucketAssigner(
+            BucketStateAndIdentifier original, BucketStateAndIdentifier updated) {
+        Cell cell = SweepAssignedBucketStoreKeyPersister.INSTANCE.sweepBucketAssignerStateMachineCell();
+        casCell(
+                cell,
+                Optional.of(bucketStateAndIdentifierPersister.trySerialize(original)),
+                bucketStateAndIdentifierPersister.trySerialize(updated));
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java:68]: please double-check: I could not find evidence that the persister is guaranteed to produce the elements in a defined order, which leads to fun times with this kind of thing. We might want to add `SORT_PROPERTIES_ALPHABETICALLY` to be 100% sure

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    private void casCell(Cell cell, Optional<byte[]> existingValue, byte[] newValue) {
+        CheckAndSetRequest request = existingValue
+                .map(value -> CheckAndSetRequest.singleCell(TABLE_REF, cell, value, newValue))
+                .orElseGet(() -> CheckAndSetRequest.newCell(TABLE_REF, cell, newValue));
+        keyValueService.checkAndSet(request);
+    }
+
+    private <T> Optional<T> readCell(Cell cell, Function<byte[], T> deserializer) {
+        Map<Cell, Value> values = keyValueService.get(TABLE_REF, ImmutableMap.of(cell, Long.MAX_VALUE));
+        System.out.println(values);
```
Comment (jeremyk-91) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java]: 👋

**Review by jeremyk-91 — COMMENTED** [review]
One thing to note for the ObjectMappers!

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java (hunk 1 above)
Comment (mdaudali) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java:68]: Yep, I'll also fix this in the bucket progress store from #7204

Comment (mdaudali) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java:68]: https://github.com/palantir/atlasdb/pull/7323!

**Review by jeremyk-91 — APPROVED** [review]
👍

**Pull request merged**